java/com/google/security/wycheproof/testcases/AesGcmTest.java

```diff
@@ -268,6 +270,169 @@ public void testByteBufferAlias() throws Exception {
+        Cipher cipher = Cipher.getInstance("AES/GCM/NoPadding");
+        cipher.init(Cipher.ENCRYPT_MODE, key, parameters);
+
+        ByteBuffer output, input, inputRO;
+
+        // We'll try three scenarios: Ordinary array backed buffers, array backed buffers where one is read-only, and
+        // direct byte buffers.
+        String mode;
+        // offsets relative to start of buffer
+        int inputOffsetInBuffer = 1;
+        int outputOffsetInBuffer = inputOffsetInBuffer + outputOffset;
+        int sliceLength = cipher.getOutputSize(ptVector.length);
+        int bufferSize = sliceLength + Math.max(inputOffsetInBuffer, outputOffsetInBuffer);
+        switch (i) {
+          case 0:
+          case 1: {
+            byte[] buffer = new byte[bufferSize];
+            // It's important to slice() here as otherwise later when we flip() position will be reset to 0.
+            output = ByteBuffer.wrap(buffer, outputOffsetInBuffer, sliceLength).slice();
+            input = ByteBuffer.wrap(buffer, inputOffsetInBuffer, sliceLength).slice();
+
+            if (i == 1) {
+              mode = "array backed buffers with RO buffer";
+              inputRO = input.asReadOnlyBuffer();
+            } else {
+              mode = "array backed buffers";
+              inputRO = input.duplicate();
+            }
+
+            break;
+          }
+          case 2: {
+            mode = "direct buffers";
+            ByteBuffer buf = ByteBuffer.allocateDirect(bufferSize);
+            output = buf.duplicate();
+            output.position(outputOffsetInBuffer);
+            output.limit(sliceLength + outputOffsetInBuffer);
+            output = output.slice();
+
+            input = buf.duplicate();
+            input.position(inputOffsetInBuffer);
+            input.limit(sliceLength + inputOffsetInBuffer);
+            input = input.slice();
+
+            inputRO = input.duplicate();
+            break;
+          }
+          default: {
+            throw new AssertionError("Unknown test index " + i);
+          }
+        }
+
+        // Now early points to all but the last byte of the buffer, and late and lateRead both point to all but the
+        // first byte of the buffer.
+        input.put(ptVector);
+        input.flip();
+        // We'll actually read via lateRead, which might be read-only
+        inputRO.limit(input.limit());
+
+        try {
+
+          int ctSize = cipher.doFinal(inputRO, output);
+
+          // Now flip the buffers around and undo everything
+          byte[] tmp = new byte[ctSize];
+          output.flip();
+          output.get(tmp);
+
+          output.clear();
+          input.clear();
+          inputRO.clear();
+
+          input.put(tmp);
+          input.flip();
+          inputRO.limit(input.limit());
+
+          cipher.init(Cipher.DECRYPT_MODE, key, parameters);
+          cipher.doFinal(inputRO, output);
+
+          output.flip();
+          assertEquals(ByteBuffer.wrap(ptVector), output);
```

Comment:
1. ByteBuffer equality is defined in terms of byte contents, and is not sensitive to the specific values of capacity/etc, except inasmuch that two bytebuffers with different `remaining()` values will be nonequal. See [the javadoc](https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#equals(java.lang.Object\)):

> Tells whether or not this buffer is equal to another object.
> 
> Two byte buffers are equal if, and only if,
> 
>     1. They have the same element type,
> 
>     2. They have the same number of remaining elements, and
> 
>     3. The two sequences of remaining elements, considered independently of their starting positions, are pointwise equal.
> 
> A byte buffer is not equal to any other type of object. 

2. Your assert succeeds because `byteBufferToHex()` performs an implicit `flip()` - which, since position() == 0 for both inputs, results in comparing two empty buffers.